src/hooks/users.ts

```diff
@@ -1,60 +1,132 @@
-import { useMemo } from "react";
-import { isLoaded } from "react-redux-firebase";
+import { useCallback, useEffect, useMemo } from "react";
+import { chunk } from "lodash";
 
-import { usersSelector, usersByIdSelector } from "utils/selectors";
-import { WithId } from "utils/id";
+import { universeUsersSelector, usersByIdSelector } from "utils/selectors";
+import { WithId, withId } from "utils/id";
 
 import { useSelector } from "hooks/useSelector";
 import { useUserLastSeenThreshold } from "hooks/useUserLastSeenThreshold";
 import { useVenueId } from "hooks/useVenueId";
-import { useSparkleFirestoreConnect } from "hooks/useSparkleFirestoreConnect";
+import {
+  useSparkleFirestoreConnect,
+  isLoaded,
+  SparkleRFQConfig,
+} from "hooks/useSparkleFirestoreConnect";
+import { useConnectRelatedVenues } from "hooks/useConnectRelatedVenues";
 
 import { User } from "types/User";
+import {
+  useFirestore,
+  useFirestoreConnect,
+  WhereOptions,
+} from "react-redux-firebase";
+import { useConnectCurrentVenueNG } from "./useConnectCurrentVenueNG";
 
-const useConnectVenueUsers = () => {
+// NOTE: Logically correct version of pulling users
+export const useConnectUniverseUsers = () => {
   const venueId = useVenueId();
+  const firestore = useFirestore();
 
-  useSparkleFirestoreConnect(
-    venueId
-      ? {
-          collection: "users",
-          where: ["enteredVenueIds", "array-contains", venueId],
-          storeAs: "users",
-        }
-      : undefined
-  );
-};
+  const { currentVenue, isCurrentVenueLoaded } = useConnectCurrentVenueNG();
+
+  const { relatedVenues, isRelatedVenuesLoaded } = useConnectRelatedVenues({
+    venueId,
+  });
+
+  // useEffect(() => {
+  //   if (!isRelatedVenuesLoaded) return;
+
+  //   const universeUserVenueIds = relatedVenues.map((venue) => venue.id);
+
+  //   const chunkedListeners = chunk(universeUserVenueIds, 10).map<
+  //     SparkleRFQConfig
+  //   >((venueIdsChunk) => ({
+  //     collection: "users",
+  //     where: ["enteredVenueIds", "array-contains", venueIdsChunk[0]],
+  //     storeAs: "universeUsers",
+  //   }));
+
+  //   firestore.setListeners(chunkedListeners);
```

Comment:
Another issue with this code as it's written here at the moment is that you're never unsetting the listeners, so even if the `venueId`'s change, you're just going to be creating more and more listeners.

This should be done with `useEffect`'s 'cleanup' pattern:

- https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

Reply:
Moreover, we will need to handle listener dups. And data merging.

Reply:
**RE: handling listener dupes:** I'm not sure that i've necessarily seen a good canonical example of one, but I would be looking into how to use a 'singleton' sort of a pattern that meshes with our hooks usage, etc.

In a past project I have done this using React context's, and while I know @thecadams had concerns with how we were using them in the past (due to potential performance issues, which i'm still not sure they actually had any impact on), but they can absolutely be used in good ways. My high level thinking here would be that a context would hold the 'singleton state' aspect of things, possibly could even have the firestore connect's in that level maybe.. and then they hooks would consume that context to know whether they needed to 'do anything', or if the singleton was already 'set up'. This is the pattern I was planning on implementing in refactoring our data to keep the firestore connect/selector logic 'together' in custom hooks.

**RE: data merging:** It depends on whether we decide we need to put the data in redux at all wouldn't it? The patterns I used in https://github.com/sparkletown/sparkle/pull/1065 returns a merged result (obviously for a different usecase though):

```typescript
const usersWhoEnteredAnyOfVenues: WithId<User>[] = await admin
    .firestore()
    .collection("users")
    .where("enteredVenueIds", "array-contains-any", venueIdsArray)
    .get()
    .then((snapshot) =>
      snapshot.docs.map((doc) => withId(doc.data() as User, doc.id))
    );
    
const authUsers: admin.auth.UserRecord[] = await Promise.all(
    chunk(usersWhoEnteredAnyOfVenues, 100).map(async (usersChunk) => {
      const usersChunkIds = usersChunk.map((user) => ({ uid: user.id }));
      const authUsersResult = await admin.auth().getUsers(usersChunkIds);

      return authUsersResult.users;
    })
  ).then((result) => result.flat());
```

Reply:
Not relevant anymore

Reply:
backlogged the above notes in https://github.com/sparkletown/internal-sparkle-issues/issues/181#issuecomment-764328990 for future reference